### PR TITLE
Bump Bazel to 2.2.0

### LINKED ISSUE #12: Bump Bazel version to 2.0.0
The current Bazel master version is 1.0.0 rc5. We would like to upgrade the build system from 0.23.2 to 1.0.0 when it released.

**Dependencise which supports Bazel >= 1.0.0** 
 - com_github_grpc_grpc: 1.2.6 supports Bazel 1.0.0
 - com_google_protobuf: Can be installed from grpc_deps
 - com_google_googleapis: Can be installed from grpc_deps
 - cython: Can be installed from grpc_deps
 - net_zlib: can be installed from grpc_deps
 - com_google_absl: lts_2019_08_08 branch works with Bazel 1.2.1
 - com_googlesource_code_cctz: current version works with Bazel 1.2.1 (b4935eef53820cf1643355bb15e013b4167a2867)
 - com_github_google_glog: current version works with Bazel 1.2.1 (ba8a9f6952d04d1403b97df24e6836227751454e)
 - com_github_gflags_gflags: current version works with Bazel 1.2.1 (28f50e0fed19872e0fd50dd23ce2ee8cd759338e)
 - com_google_googletest: current version works with Bazel 1.2.1 (3525e3984282c827c7207245b1d4a47f4eaf3c91)
 - com_googlesource_code_re2: current version works with Bazel 1.2.1 (be0e1305d264b2cbe1d35db66b8c5107fc2a727e)
 - com_github_opennetworkinglab_fabric_p4test: should be able to support new Bazel version since it only has a single file-group

**Will support soon**
 - io_bazel_rules_python: 0.0.2 supports Bazel 2.0

**Build file need to be updated**
 - com_github_p4lang_p4c: We are using our own build file
 - com_github_p4lang_p4runtime: We are using our own build file
 - com_github_openconfig_gnmi_proto: We are using our own build file
 - com_github_openconfig_gnoi: We are using our own build file
 - com_github_openconfig_public: We are using our own build file
 - com_github_openconfig_hercules: We are using our own build file
 - com_github_yang_models_yang: We are using our own build file
 - com_github_systemd_systemd: We are using our own build file
 - com_github_opennetworkinglab_sdklt: we are using our own build file

**Doesn't support, and there is no plan yet**
 - judy: build file from PI, not support
 - build_stack_rules_proto: Doesn't support 1.0.0
 - com_github_p4lang_PI: Doesn't support 1.0.0
 - com_github_p4lang_PI_bf: Doesn't support 1.0.0

**Can be removed**
 - boringssl: chromium-stable-with-bazel branch doesn't support it, but can be removed since no one is depend on it

**Need to be upgraded**
 - com_github_nelhage_rules_boost: Need to bump to latest version
 - rules_cc: need to bump to latest version
 - bazel_latex: need to upgrade to v0.19 (tested with Bazel 2.0.0)
 - io_bazel_rules_go: bump to latest release (v0.20.3)
 - bazel_gazelle: bump to v0.19.1

### PR DESCRIPTION
**Pull request by Yi-Tseng**
Fix #12

### CODE REVIEW

**Comment by pudelkoM** [comment]
I'd suggest moving directly to 2.0.0

**Comment by pudelkoM** [comment]
Good news! Most of our targets still build with 2.0.0. E.g. `stratum_bcm` and the `/hal` tree.
Only the P4c related targets fail because of a boost dependency:

**Comment by devjitgopalpur** [comment]
> Good news! Most of our targets still build with 2.0.0. E.g. `stratum_bcm` and the `/hal` tree.
> Only the P4c related targets fail because of a boost dependency:

Is this PR bumping version to 2.0.0?

**Comment by pudelkoM** [comment]
> > Good news! Most of our targets still build with 2.0.0. E.g. `stratum_bcm` and the `/hal` tree.
> > Only the P4c related targets fail because of a boost dependency:
> 
> Is this PR bumping version to 2.0.0?

Not yet. @Yi-Tseng started with 1.X, but I'd tried 2.0.0 because why should we be stuck on an old version again?

**Comment by pudelkoM** [comment]
All `test-targets.txt` build and pass!

**Comment by pudelkoM** [comment]
All build and tests targets work now. I think we made it!

File: stratum/BUILD [hunk 1]
```diff
@@ -17,12 +17,10 @@
...
-load(
-    "//bazel:rules.bzl",
-    "STRATUM_INTERNAL",
-    "stratum_package",
-    "stratum_platform_select",
-)
+load("//bazel/rules:platform_rules.bzl",
+     "stratum_platform_select", "stratum_platform_filter", "stratum_platform_alias")
+load("//bazel/rules:package_rule.bzl", "stratum_package")
+load("//bazel:defs.bzl", "EMBEDDED_ARCHES", "HOST_ARCHES", "STRATUM_INTERNAL")
```
Comment (pudelkoM) [inline stratum/BUILD]: This can be reverted back.

File: stratum/glue/BUILD [hunk 2]
```diff
@@ -17,11 +17,8 @@
 
 licenses(["notice"])  # Apache v2
 
-load(
-    "//bazel:rules.bzl",
-    "stratum_cc_library",
-    "STRATUM_INTERNAL",
-)
+load("//bazel:defs.bzl", "STRATUM_INTERNAL")
```
Comment (pudelkoM) [inline stratum/glue/BUILD]: same here

File: stratum/glue/net_util/BUILD [hunk 3]
```diff
@@ -17,12 +17,9 @@
 
 licenses(["notice"])  # Apache v2
 
-load(
-    "//bazel:rules.bzl",
-    "STRATUM_INTERNAL",
-    "stratum_cc_library",
-    "stratum_cc_test",
-)
+load("//bazel/rules:library_rule.bzl", "stratum_cc_library")
```
Comment (pudelkoM) [inline stratum/glue/net_util/BUILD]: same here

File: stratum/glue/status/BUILD [hunk 4]
```diff
@@ -17,12 +17,9 @@
 
 licenses(["notice"])  # Apache v2
 
-load(
-    "//bazel:rules.bzl",
-    "STRATUM_INTERNAL",
-    "stratum_cc_library",
-    "stratum_cc_test",
-)
+load("//bazel/rules:library_rule.bzl", "stratum_cc_library")
```
Comment (pudelkoM) [inline stratum/glue/status/BUILD]: same

File: stratum/hal/config/BUILD [hunk 5]
```diff
@@ -15,12 +15,10 @@
 # limitations under the License.
 #
 
-load(
-    "//bazel:rules.bzl",
-    "STRATUM_INTERNAL",
-    "stratum_package",
-    "stratum_cc_test",
-)
+load("//bazel/rules:test_rule.bzl", "stratum_cc_test")
```
Comment (pudelkoM) [inline stratum/hal/config/BUILD]: same

**Comment by bocon13** [comment]
> I'd suggest moving directly to 2.0.0.
> 
> We could also think about dropping some of our special rules (`stratum_cc_library`)
> and use the official ones.

@devjitgopalpur One of the reasons that we kept these is so that someone else (e.g. Google) could use a different implementation (e.g. compiler options/toolchain/etc.); is this still a use case that you care about?

**Comment by devjitgopalpur** [comment]
> > We could also think about dropping some of our special rules (`stratum_cc_library`)
> > and use the official ones.
> 
> @devjitgopalpur One of the reasons that we kept these is so that someone else (e.g. Google) could use a different implementation (e.g. compiler options/toolchain/etc.); is this still a use case that you care about?

Yes, I think we might want to keep these around. Would be handy for compiling to another architecture.

File: WORKSPACE [hunk 6]
```diff
@@ -35,15 +35,18 @@ workspace(name = "com_github_stratum_stratum")
 #       Please do not push changes to this section upstream.
 # ---------------------------------------------------------------------------
 
+#local_repository(
+#    name = "com_github_grpc_grpc",
+#    path = "/grpc-1.26.0"
+#)
+
```
Comment (Yi-Tseng) [inline WORKSPACE]: to be removred

File: bazel/patches/grpc.patch [hunk 7]
```diff
@@ -1,124 +1,24 @@
-From 76c126b8f7d35134be88165c63bec872c88e58f2 Mon Sep 17 00:00:00 2001
```
Comment (Yi-Tseng) [inline bazel/patches/grpc.patch]: this file can be removed.

**Comment by Yi-Tseng** [comment]
> All build and tests targets we check in CI work now. I think we made it!
> 
> @Yi-Tseng I'm not sure about the code chunks you commented out and leave them to you.

I am still testing python-grpc targets.
Those targets are:
P4Runtime gRPC python: for PTF
gNMI gRPC python: for gNMI CLI
dummy gRPC library: for dummy CLI

We can remove gNMI gRPC python library if we want to move to go/c++ version gNMI CLI (#74)
we can also remove dummy gRPC python library since we can re-implement the dummy CLI with c++

**Comment by pudelkoM** [comment]
> > > We could also think about dropping some of our special rules (`stratum_cc_library`)
> > > and use the official ones.
> > 
> > 
> > @devjitgopalpur One of the reasons that we kept these is so that someone else (e.g. Google) could use a different implementation (e.g. compiler options/toolchain/etc.); is this still a use case that you care about?
> 
> Yes, I think we might want to keep these around. Would be handy for compiling to another architecture.

Sure. They didn't cause any issues anyway.

File: .circleci/config.yml [hunk 8]
```diff
@@ -78,7 +78,7 @@ jobs:
   # Build targets and run unit tests.
   unit_tests:
     docker:
-      - image: stratumproject/build:build
+      - image: stratumproject/build:bazel-2.0.0
```
Comment (Yi-Tseng) [inline .circleci/config.yml]: Will change back after merged

**Comment by pudelkoM** [comment]
CI fails because of this bazel [bug](https://github.com/bazelbuild/bazel/issues/10140). It's just `analyze-profile`, so not really important. We could either temporarily disable the command or remove it.

**Comment by meisterT** [comment]
I am working on the said Bazel issue and will submit basic support for the JSON profile soon.
However, instead of creating the HTML page with the help of analyze-profile, you should just upload and store the JSON profile file itself. You can then later load it into `chrome://tracing` (no need to create an HTML page as intermediate step).

If you really want to create the HTML page still, then you can follow the instructions in https://github.com/catapult-project/catapult/tree/master/tracing to do so.

File: bazel/deps.bzl [hunk 9]
```diff
@@ -217,27 +214,14 @@ def stratum_deps():
...
-        )
-
-    if "rules_cc" not in native.existing_rules():
-        remote_workspace(
-            name = "rules_cc",
-            remote = "https://github.com/bazelbuild/rules_cc",
-            commit = "cfe68f6bc79dea602f2f6a767797f94a5904997f",
+            shallow_since = "1570056263 -0700",
         )
+        # sha256 = "bd3155d1f13792a798e116034322dbe2ee2904253396573ab64ab10dc8c27d96"
```
Comment (pudelkoM) [inline bazel/deps.bzl]: Remove?

**Comment by pudelkoM** [comment]
> I am working on the said Bazel issue and will submit basic support for the JSON profile soon.
> However, instead of creating the HTML page with the help of analyze-profile, you should just upload and store the JSON profile file itself. You can then later load it into `chrome://tracing` (no need to create an HTML page as intermediate step).
> 
> If you really want to create the HTML page still, then you can follow the instructions in https://github.com/catapult-project/catapult/tree/master/tracing to do so.

I think what we care most about is this text output:
```
=== PHASE SUMMARY INFORMATION ===

Total launch phase time          978 ms    2.67%
Total init phase time            178 ms    0.49%
Total loading phase time         381 ms    1.04%
Total analysis phase time       2.618 s    7.15%
Total preparation phase time    46.0 ms    0.13%
Total execution phase time     32.366 s   88.46%
Total finish phase time         19.3 ms    0.05%
Total run time                 36.586 s  100.00%
```

That the details can be looked at in Chrome is an acceptable substitute for the HTML page. (Apart from Chrome...)

File: .circleci/config.yml [hunk 10]
```diff
@@ -96,9 +96,12 @@ jobs:
   # Generate coverage report and upload to codecov.io.
   coverage:
     docker:
-      - image: stratumproject/build:build
+      - image: stratumproject/build:bazel-2.0.0
     steps:
       - checkout
+      - run:
+          name: add gcov wrapper
+          command: rm /usr/bin/gcov && cp .circleci/gcov-wrapper.sh /usr/bin/gcov
```
Comment (Yi-Tseng) [inline .circleci/config.yml]: maybe merge to base docker image?

File: .circleci/test-targets.txt [hunk 11]
```diff
@@ -1,13 +1,10 @@
 //stratum/glue/gtl/...
 //stratum/glue/net_util/...
 //stratum/glue/status/...
-//stratum/hal/bin/bcm/standalone/...
-//stratum/hal/lib/bcm/...
```
Comment (Yi-Tseng) [inline .circleci/test-targets.txt]: add this back after CI pass and store caches, same as below

File: .circleci/config.yml [hunk 12]
```diff
@@ -36,18 +36,17 @@ save_bazel_cache: &save_bazel_cache
     paths:
       - /tmp/bazel-cache
 
-# Print Bazel profiling info and generate HTML report to upload as artifact.
+# Wait until JSON profile analyzer released
+# See: https://github.com/bazelbuild/bazel/issues/10140
```
Comment (pudelkoM) [inline .circleci/config.yml]: I think this comment is obsolete now.

Comment (pudelkoM) [inline .circleci/config.yml]:
```suggestion
# Print Bazel profiling info and generate JSON report to upload as artifact.
```

File: stratum/hal/lib/phal/BUILD [hunk 13]
```diff
@@ -20,8 +20,8 @@ licenses(["notice"])  # Apache v2
 load(
     "//bazel:rules.bzl",
     "STRATUM_INTERNAL",
-    "stratum_cc_binary",
     "stratum_cc_library",
+    "stratum_cc_binary",
```
Comment (pudelkoM) [inline stratum/hal/lib/phal/BUILD]: Resort

File: stratum/hal/lib/common/admin_service.h [hunk 14]
```diff
@@ -33,6 +32,7 @@
 #include "stratum/hal/lib/common/switch_interface.h"
 #include "stratum/lib/security/auth_policy_checker.h"
 #include "stratum/lib/timer_daemon.h"
+#include "gnoi/system/system.grpc.pb.h"
```
Comment (pudelkoM) [inline stratum/hal/lib/common/admin_service.h]: Move to top again

**Pull request merged**